src/grt/grt-rtiis.adb

```diff
@@ -0,0 +1,350 @@
+            begin
+               Layout_Addr := Loc_To_Addr (
+                  St.Common.Depth, St.Range_Loc, Ctxt);
+            end;
+         when Ghdl_Rtik_Type_I32
+           | Ghdl_Rtik_Type_E8
+           | Ghdl_Rtik_Type_E32
+           | Ghdl_Rtik_Type_B1 =>
+            Layout_Addr := Null_Address;
+         when Ghdl_Rtik_Type_Array =>
+            -- We can't have an unbounded type as the port/signal.
+            -- and there's nowhere for the Rtin_Obj to hold the bounds.
+            Internal_Error("to_ghdl_rtii");
+         when Ghdl_Rtik_Subtype_Array =>
+            declare
+               St : constant Ghdl_Rtin_Subtype_Composite_Acc :=
+                 To_Ghdl_Rtin_Subtype_Composite_Acc (Rti);
+            begin
+               Layout_Addr := Loc_To_Addr (St.Common.Depth, St.Layout, Ctxt);
+            end;
+         when Ghdl_Rtik_Type_Record =>
+            declare
+               Bt : constant Ghdl_Rtin_Type_Record_Acc :=
+                 To_Ghdl_Rtin_Type_Record_Acc (Rti);
+            begin
+               if Rti_Complex_Type (Rti) then
+                  Layout_Addr := Loc_To_Addr (
+                     Bt.Common.Depth, Bt.Layout, Ctxt);
+               else
+                  -- FIXME:  Is this right?
+                  --Layout_Addr := Bt.Layout;
+                  null;
+               end if;
+            end;
+         when Ghdl_Rtik_Subtype_Record =>
+            declare
+               St : constant Ghdl_Rtin_Subtype_Composite_Acc :=
+                 To_Ghdl_Rtin_Subtype_Composite_Acc (Rti);
+            begin
+               Layout_Addr := Loc_To_Addr (St.Common.Depth, St.Layout, Ctxt);
+            end;
+         when others =>
+            Internal_Error("to_ghdl_rtii");
+      end case;
+      return Layout_Addr;
+   end Get_Type_Layout;
+
+   function Get_Array_Nbr_Children(Rtii : Ghdl_Rtii) return Ghdl_Index_Type is
+      Kind : Ghdl_Rtik;
+      Base_Array : Ghdl_Rtin_Type_Array_Acc;
+      Layout_Addr : Address;
+      Bounds_Addr : Address;
+      Length : Ghdl_Index_Type;
+      Idx_Rti : Ghdl_Rti_Access;
+      Rng : Ghdl_Range_Ptr;
+   begin
+      Kind := Rtii.Typ.Rti.Kind;
+      case Kind is
+         when Ghdl_Rtik_Type_Array =>
+            Base_Array := To_Ghdl_Rtin_Type_Array_Acc(Rtii.Typ.Rti);
+            Layout_Addr := Rtii.Typ.Binding_Layout_Addr;
+         when Ghdl_Rtik_Subtype_Array =>
+            Base_Array := To_Ghdl_Rtin_Type_Array_Acc(
+               To_Ghdl_Rtin_Subtype_Composite_Acc(Rtii.Typ.Rti).Basetype);
+            Layout_Addr := Rtii.Typ.Layout_Addr;
+         when others =>
+            Internal_Error("get_array_element_type");
+      end case;
+      Bounds_Addr := Layout_Addr + Ghdl_Index_Type'(
+         Ghdl_Indexes_Type'Size / 8);
+      -- Get the length of this dimension so we can check whether
+      -- the index is valid.
+      Idx_Rti := Get_Base_Type (Base_Array.Indexes (Rtii.Typ.Dim));
+      Extract_Range (Bounds_Addr, Idx_Rti, Rng);
```

Comment:
Maybe the multi-dim arrays should be flattened (as if they had only one dimension).  That could simplify the code.

Reply:
Yep.  I need to check how this is handled in VPI and VHPI and then try to stay close to that.